### PR TITLE
feat(Tactic/Linter/Style): add a linter for indenting of declarations

### PR DESCRIPTION
**Pull request by Hagb**
to check whether type and `:=` of a declaration are correctly indented as described in
https://leanprover-community.github.io/contribute/style.html#structuring-definitions-and-theorems.

---

- [ ] depends on: #27502

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by grunweg** [comment]
Thanks for writing this linter, this is great! I got very excited and made a PR #27478 fixing all violations in the first 1000 files (of the build log above).

**Comment by grunweg** [comment]
I hope this does not duplicate any work you did (I'm certainly happy to let you go ahead with any changes you made).
That said, there are 6000 files left, so there should be enough work for everybody.

**Comment by Hagb** [comment]
> I hope this does not duplicate any work you did (I'm certainly happy to let you go ahead with any changes you made). That said, there are 6000 files left, so there should be enough work for everybody.

Thank you! Feel free to submit these fixes!
By the way I think it can be fixed more easily with warnings filtered from the log. One can open it (`vi`, `less`, etc.) in the terminal of VSCode, and VSCode can jump to the location when ctrl+click is pressed on a line.

I got it when I was debugging the linter:
[warning.txt](https://github.com/user-attachments/files/21436400/warning.txt)

(There are 300+ lines to be fixed.)

```bash
lake build | tee build.log # or save the build log from the workflow
grep warning build.log > warning.txt
```

File: Mathlib/Tactic/Linter/Style.lean [hunk 1]
```diff
@@ -10,6 +10,8 @@ import Lean.Server.InfoUtils
 -- this file has a valid copyright header and module docstring.
 import Mathlib.Tactic.Linter.Header
 import Mathlib.Tactic.DeclarationNames
+import Batteries.Data.Array
+import Qq
```
Comment (adomani) [inline Mathlib/Tactic/Linter/Style.lean]: Are these imports really needed?  Especially `Qq` can probably just be removed.  I'm also not too convinced about adding the `Batteries` dependendency, but I'm less worried by that.

**Comment by grunweg** [comment]
Thanks! Let me fix another batch of warnings; I'll comment when I'm done.

File: Mathlib/Tactic/Linter/Style.lean (hunk 1 above)
Comment (Hagb) [reply Mathlib/Tactic/Linter/Style.lean]: > Especially `Qq` can probably just be removed.

Yeah. Thank you! I forgot to check it after I copied it from my draft.

> I'm also not too convinced about adding the `Batteries` dependendency

I imported it for [`Array.mergeDedupwith`](https://leanprover-community.github.io/mathlib4_docs/Batteries/Data/Array/Merge.html#Array.mergeDedupWith`) to merge two sorted `Array Range` (denoted `Ranges` in the PR) and also merge overlapped or adjacent `Range`s, in `mergeRanges`.

Is it okay to just copy the code of [`Array.mergeDedupwith`](https://leanprover-community.github.io/mathlib4_docs/Batteries/Data/Array/Merge.html#Array.mergeDedupWith`) from Batteries? This function itself is independent of other stuff in Batteries.
And it seems to me that practically there should be just few elements in such an array, so I think it doesn't hurt to even implement it with a slower but simpler way (if we don't want to copy code from Batteries or reimplement it)?

**Comment by grunweg** [comment]
I'll release the "lock" on this PR for now.

**Comment by grunweg** [comment]
I'm working on some more fixes now.

**Comment by grunweg** [comment]
Fixing the linter violations, I realised that the current implementation does not check further indentation at all. Is this on purpose or an oversight? (I can imagine it being difficult in the general case: you want to allow additional indentation, and sometimes by an odd number of spaces to deal with angle brackets. Maybe you can get away with some heuristic such as linting an *odd* number of spaces, if there is no comma or angle bracket in the statement.)
Can you add tests for this, either? Tests for the tricky cases are particularly great. Mathlib should provide plenty of inspiration :-)

**Comment by grunweg** [comment]
Another question: how difficult is to check that the proofs are indented by two spaces, not more? (Perhaps a good check is that *some part* of the proof is two-space indented. Not thinking about further indentation for sub-goals, `have`s, line breaks etc. seems like good enough for a first version.)
(edit: this can and perhaps should be a follow-up PR to this linter --- but would also be really nice to have!)

**Comment by grunweg** [comment]
I'm done for tonight.

File: Mathlib/Tactic/Linter/Style.lean (hunk 1 above)
Comment (grunweg) [reply Mathlib/Tactic/Linter/Style.lean]: If you're just importing a single file from batteries, perhaps it's fine to just do so.

**Comment by grunweg** [comment]
#27502 has perhaps the last indentation fixes to make mathlib's `master` comply with the linter.

**Comment by grunweg** [comment]
Sleeping over this, I think the following checks would be worthwhile. They are fairly conservative, so hopefully have almost no false positives. Each of these could happen in a future PR; I'd rather land this PR now than delay it over these enhancements.
- the first line of the declaration (i.e, the `lemma`, `theorem` statement, etc.) is not indented (you implement this already, right?)
- if there are further lines, some of these is indented by four spaces
  (in other words, indenting everything by *six* spaces should give an error)
  We don't check more strictly to give maximum flexibility with line breaks.
- checking for lines indented by an odd number of spaces can happen in the future

In a proof body,
- every line is indented by at least two spaces
  (stronger version: proofs of a have, suffices, etc. are indented by at least two spaces more than the `have`/`suffices` itself)
- some line is indented by exactly two spaces
  i.e., indenting everything by four spaces is caught
  Lean will check most of the remaining consistency issues: for instance, a proof
  ```
  simp
    exact foo
  ```
  (which is meant to be equivalent to `simp; exact foo`) will already error.


Another check, similar but conceptually independent of this PR: if there is just a second line for the statement and the proof is longer than X (20 characters? 30 characters); it should be on a separate line
```
instance of_hasKernel_of_hasCokernel [HasKernel S.g] [HasCokernel (kernel.lift S.g S.f S.zero)] :
    S.HasLeftHomology := HasLeftHomology.mk' (LeftHomologyData.ofHasKernelOfHasCokernel S)
```
should be flagged

Another check: `where` (in a structure definition) should be on a separate line ---
```
noncomputable instance colimitOne : One (M.{v, u} F) where one := M.mk F ⟨IsFiltered.nonempty.some,1⟩
```
should be linted

**Comment by grunweg** [comment]
All of these are very clearly future enhancements of this PR. In other words, I'm very excited about your linter!

**Comment by grunweg** [comment]
I have filed #27505 to record them.

**Comment by Hagb** [comment]
@grunweg Thank you for your work! I will merge them into this PR later to pass the workflow.

> you want to allow additional indentation, and sometimes by an odd number of spaces to deal with angle bracket

Is angle bracket an exception where odd number of spaces are preferred or at least not bad? After searching in mathlib, I found some samples like
https://github.com/leanprover-community/mathlib4/blob/2cb345fe38f9954ef54ca97809f1e304f2dcac8a/Mathlib/CategoryTheory/Iso.lean#L139-L144 where `⟨` is indented by 5 spaces.

**Comment by Hagb** [comment]
It can now check indenting of modifiers and the first line of declaration, while disabled by default:
```lean
import Mathlib.Tactic.Linter.Style

set_option linter.style.declarationIndenting true
set_option linter.style.declarationIndenting.modifiersAndFirstLine true

/-- docstring -/
  @[simp] -- warning: Here should not be spaces at the beginning of this line
noncomputable -- ok, no warning
  example : -- warning: Here should not be spaces at the beginning of this line
True := trivial -- warning: There are too few spaces at the beginning of the line. It should be indented by at least 4 spaces.
```

**Comment by grunweg** [comment]
> @grunweg Thank you for your work! I will merge them into this PR later to pass the workflow.
> 
> > you want to allow additional indentation, and sometimes by an odd number of spaces to deal with angle bracket
> 
> Is angle bracket an exception where odd number of spaces are preferred or at least not bad? After searching in mathlib, I found some samples like
> 
> https://github.com/leanprover-community/mathlib4/blob/2cb345fe38f9954ef54ca97809f1e304f2dcac8a/Mathlib/CategoryTheory/Iso.lean#L139-L144
> where `⟨` is indented by 5 spaces.

I think that one should be 6 spaces.

I was thinking of "if the term in angle brackets span multiple lines, indenting subsequent lines by one space could be fine" --- i.e., be less strict about the *next* lines. The line containing the opening angle bracket should (probably) always be indented by an even number of spaces.

**Comment by grunweg** [comment]
By the way: can you add tests for all these edge cases? These are extremely valuable to have, as the linter is developed over time.

**Comment by grunweg** [comment]
(This PR is labelled WIP. If you think this is ready for review, please unlabel or say that you're looking for review. Otherwise, I'll assume you still want to polish this more.)

**Comment by Hagb** [comment]
> By the way: can you add tests for all these edge cases? These are extremely valuable to have, as the linter is developed over time.

Yes. I will add them (maybe to #27525), while linters for some of them haven't been implemented.

> (This PR is labelled WIP. If you think this is ready for review, please unlabel or say that you're looking for review. Otherwise, I'll assume you still want to polish this more.)

I am actually thinking of closing this one, since it has (almost) been reimplemented/refactored as a (much shorter) component in the framework of #27525.

**Comment by grunweg** [comment]
If this linter is much shorter as part of #27525, then I'm happy to just review that one instead.
